cdap-app-fabric/src/main/java/io/cdap/cdap/internal/capability/CapabilityApplier.java

```diff
@@ -0,0 +1,374 @@
+        LOG.debug("Capability {} config for status {} skipped because there is already an operation {} in progress.",
+                  capability, newConfig.getStatus(), currentOperations.get(capability).getActionType());
+        continue;
+      }
+      switch (newConfig.getStatus()) {
+        case ENABLED:
+          enableSet.add(newConfig);
+          break;
+        case DISABLED:
+          disableSet.add(newConfig);
+          break;
+        default:
+          break;
+      }
+      currentCapabilities.remove(capability);
+    }
+    //add all unfinished operations to retry
+    for (CapabilityOperationRecord operationRecord : currentOperations.values()) {
+      switch (operationRecord.getActionType()) {
+        case ENABLE:
+          enableSet.add(operationRecord.getCapabilityConfig());
+          break;
+        case DISABLE:
+          disableSet.add(operationRecord.getCapabilityConfig());
+          break;
+        case DELETE:
+          deleteSet.add(operationRecord.getCapabilityConfig());
+          break;
+        default:
+          break;
+      }
+      currentCapabilities.remove(operationRecord.getCapability());
+    }
+    // find the ones that are not being applied or retried - these should be removed
+    deleteSet.addAll(currentCapabilities.values().stream()
+                       .map(CapabilityStatusRecord::getCapabilityConfig).collect(Collectors.toSet()));
+    enableCapabilities(enableSet);
+    disableCapabilities(disableSet);
+    deleteCapabilities(deleteSet);
+  }
+
+  private void enableCapabilities(Set<CapabilityConfig> enableSet) throws Exception {
+    Map<ProgramId, Arguments> enabledPrograms = new HashMap<>();
+    for (CapabilityConfig capabilityConfig : enableSet) {
+      //collect the enabled programs
+      capabilityConfig.getPrograms().forEach(systemProgram -> enabledPrograms
+        .put(getProgramId(systemProgram), new BasicArguments(systemProgram.getArgs())));
+      String capability = capabilityConfig.getCapability();
+      CapabilityConfig existingConfig = capabilityStatusStore.getConfig(capability);
+      if (capabilityConfig.equals(existingConfig)) {
+        capabilityStatusStore.deleteCapabilityOperation(capability);
+        continue;
+      }
+      capabilityStatusStore.addOrUpdateCapabilityOperation(capability, CapabilityAction.ENABLE, capabilityConfig);
+      LOG.debug("Enabling capability {}", capability);
+      //If already deployed, will be ignored
+      deployAllSystemApps(capability, capabilityConfig.getApplications());
+    }
+    //start all programs
+    systemProgramManagementService.setProgramsEnabled(enabledPrograms);
+    //mark all as enabled
+    for (CapabilityConfig capabilityConfig : enableSet) {
+      String capability = capabilityConfig.getCapability();
+      capabilityStatusStore
+        .addOrUpdateCapability(capability, CapabilityStatus.ENABLED, capabilityConfig);
+      capabilityStatusStore.deleteCapabilityOperation(capability);
```

Comment:
So this call and the previous call are in different transaction, is it ok? E.g. if the status changed to ENABLED, but the operation was not deleted.

Reply:
Should not be a problem. Next attempt, this will be noticed as ongoing operation , but since the config is same , action will not be applied and the operation entry will get deleted.